File: python/rtc/world.py
import math

from rtc.color import Color
from rtc.intersection import Intersection, Intersections, Computations
from rtc.lights import PointLight, Light
from rtc.ray import Ray
from rtc.shape import Shape
from rtc.sphere import Sphere
from rtc.transform import *
from rtc.tuples import *

from dataclasses import dataclass
from typing import List

out_ray = Ray(Point(0, 0, 0), Vector(0, 0, 0))
RAY_BOUNCE_DEPTH = 4


@dataclass
class World:
    shapes: List["Shape"]
    lights: List["Light"]

    def intersect(self, ray: "Ray") -> "Intersections":
        intersections: list[Intersection] = []
        for o in self.shapes:
            for i in o.intersect(ray, out_ray):
                intersections.append(i)

        return Intersections(sorted(intersections, key=lambda i: i.t))

    def shade_hit(self, comps: "Computations", remaining: int = RAY_BOUNCE_DEPTH) -> "Color":
        out = Color(0, 0, 0)
        for light in self.lights:
            material = comps.shape.material
            comps.point
            eyev = comps.eyev
            normalv = comps.normalv
            shape = comps.shape

            shadowed = self.is_shadowed(comps.over_point)

            surface = material.lighting(
                shape, light, comps.over_point, eyev, normalv, shadowed
            )
            reflected = self.reflected_color(comps, remaining)
            refracted = self.refracted_color(comps, remaining)

            if material.reflective > 0 and material.transparency > 0:
                reflectance = comps.schlick()
                out += surface + reflected * reflectance + refracted * (1 - reflectance)
                continue

            out += surface + reflected + refracted
        return out

    def color_at(self, ray: "Ray", remaining: int = RAY_BOUNCE_DEPTH) -> "Color":
        intersections = self.intersect(ray)
        hit = intersections.hit()

        if hit is None:
            return Color(0, 0, 0)

        comps = hit.prepare_computations(ray, intersections)
        return self.shade_hit(comps, remaining)

    def is_shadowed(self, point: "Tuple4") -> bool:
        for light in self.lights:
            v = tuple4_sub(light.position, point)
            distance = tuple4_magnitude(v)
            direction = tuple4_normalize(v)

            r = Ray(point, direction)
            intersections = self.intersect(r)

            h = intersections.hit(shadow=True)

            if h is not None and h.t < distance:
                return True

        return False

    def reflected_color(self, comps: "Computations", remaining: int = RAY_BOUNCE_DEPTH) -> "Color":
        reflective = comps.shape.material.reflective
        if reflective == 0 or remaining < 1:
            return Color(0, 0, 0)

        reflect_ray = Ray(comps.over_point, comps.reflectv)
        color = self.color_at(reflect_ray, remaining - 1)
        return color * reflective

    def refracted_color(self, comps, remaining: int = RAY_BOUNCE_DEPTH):
        if remaining == 0:
            return Color(0, 0, 0)
        if comps.shape.material.transparency == 0:
            return Color(0, 0, 0)

        n_ratio = comps.n1 / comps.n2
        cos_i = tuple4_dot(comps.eyev, comps.normalv)
        sin2_t = (n_ratio * n_ratio) * (1 - (cos_i * cos_i))
        if sin2_t > 1:
            return Color(0, 0, 0)

        cos_t = math.sqrt(1 - sin2_t)
        direction = tuple4_sub(
            tuple4_scale(comps.normalv, (n_ratio * cos_i - cos_t)),
            tuple4_scale(comps.eyev, n_ratio),
        )
        reflect_ray = Ray(comps.under_point, direction)
        color = (
            self.color_at(reflect_ray, remaining - 1)
            * comps.shape.material.transparency
        )

        return color


def DefaultWorld() -> "World":
    light = PointLight(Point(-10, 10, -10), Color(1, 1, 1))
    s1 = Sphere()
    s1.material.color = Color(0.8, 1.0, 0.6)
    s1.material.diffuse = 0.7
    s1.material.specular = 0.2

    s2 = Sphere()
    s2.transform = scaling(0.5, 0.5, 0.5)

    return World([s1, s2], [light])

File: c/world.c

#include "labrat.h"

#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#include "color.h"
#include "error.h"
#include "ray.h"
#include "shape.h"
#include "transform.h"
#include "vec4.h"
#include "world.h"
#include "util.h"

world_t world() {
	world_t w;

	w.n_point_lights = 0;
	w.cap_point_lights = 16;
	w.point_lights = (point_light_t*)calloc(w.cap_point_lights, sizeof(point_light_t));

	w.n_shapes = 0;
	w.cap_shapes = 16;
	w.shapes = (shape_t*)calloc(w.cap_shapes, sizeof(shape_t));

	return w;
}

void world_add_point_light(world_t *w, point_light_t pl) {
	if (w->n_point_lights >= w->cap_point_lights - 1) {
		int oc = w->cap_point_lights;
		w->cap_point_lights *= 2;
		w->point_lights = (point_light_t*)realloc(
				w->point_lights, sizeof(point_light_t)*w->cap_point_lights);
		if (w->point_lights == NULL) {
			fprintf(stderr, "Unable to grow world point_lights from %d to %d", oc, w->cap_point_lights);
			exit(ERROR_GROW_WORLD_POINT_LIGHTS_LIST);
			return;
		}
	}

	w->point_lights[w->n_point_lights++] = pl;
}

void world_add_shape(world_t *w, shape_t s) {
	if (w->n_shapes >= w->cap_shapes - 1) {
		int oc = w->cap_shapes;
		w->cap_shapes *= 2;
		w->shapes = (shape_t*)realloc(
				w->shapes, sizeof(shape_t)*w->cap_shapes);
		if (w->shapes == NULL) {
			fprintf(stderr, "Unable to grow world shapes from %d to %d", oc, w->cap_shapes);
			exit(ERROR_GROW_WORLD_POINT_LIGHTS_LIST);
			return;
		}
	}

	w->shapes[w->n_shapes++] = s;
}

void world_free(world_t w) {
	free(w.point_lights);
	for (int i = 0; i < w.n_shapes; i++) {
		shape_free(w.shapes[i]);
	}
	free(w.shapes);
}

world_t default_world() {
	world_t w = world();
	world_add_point_light(&w, point_light(point(-10, 10, -10), color(1, 1, 1)));

	shape_t s1 = sphere();
	s1.material.color = color(0.8, 1, 0.6);
	s1.material.diffuse = 0.7;
	s1.material.specular = 0.2;
	world_add_shape(&w, s1);

	shape_t s2 = sphere();
	shape_set_transform(&s2, scaling(0.5, 0.5, 0.5));
	world_add_shape(&w, s2);

	return w;
}

intersection_list_t world_intersect_ray(world_t w, ray_t r) {
	intersection_list_t ilist = intersection_list();

	for (int i = 0; i < w.n_shapes; i++) {
		w.shapes[i].intersect(w.shapes[i], r, &ilist);
	}

	return ilist;
}

color_t world_shade_hit(world_t w, computations_t c) {
	color_t res = color(0, 0, 0);

	for (int i = 0; i < w.n_point_lights; i++) {
		bool is_shadowed = world_point_is_shadowed(w, c.over_point, i);
		color_t hit = lighting(
					c.s.material,
					w.point_lights[i],
					c.over_point,
					c.eyev,
					c.normalv,
					is_shadowed
					);
		res = color_add(res, hit);
	}

	return res;
}

color_t world_color_at(world_t w, ray_t r) {
	intersection_list_t ilist = world_intersect_ray(w, r);
	if (ilist.hit == NULL) {
		intersection_list_free(ilist);
		return color(0, 0, 0);
	}

	computations_t comps = prepare_computations(*ilist.hit, r);
	intersection_list_free(ilist);
	return world_shade_hit(w, comps);
}

bool world_point_is_shadowed(world_t w, vec4_t p, int point_lights_i) {
	vec4_t v = vec4_sub(w.point_lights[point_lights_i].position, p);
	double distance = vec4_magnitude(v);
	vec4_t direction = vec4_normalize(v);

	ray_t r = ray(p, direction);
	intersection_list_t ilist = world_intersect_ray(w, r);

	bool res = ilist.hit != NULL && ilist.hit->t < distance;
	intersection_list_free(ilist);
	return res;
}

TEST_CASE(creating_a_world) {
	world_t w = world();
	ASSERT_TRUE(w.n_point_lights == 0);
	ASSERT_TRUE(w.n_shapes == 0);
	world_free(w);
}

TEST_CASE(the_default_world) {
	point_light_t light = point_light(point(-10, 10, -10), color(1, 1, 1));

	shape_t s1 = sphere();
	s1.material.color = color(0.8, 1, 0.6);
	s1.material.diffuse = 0.7;
	s1.material.specular = 0.2;

	shape_t s2 = sphere();
	shape_set_transform(&s2, scaling(0.5, 0.5, 0.5));

	world_t w = default_world();
	ASSERT_TRUE(w.n_point_lights == 1);
	ASSERT_TRUE(w.n_shapes == 2);

	ASSERT_TRUE(point_light_eq(w.point_lights[0], light));
	ASSERT_TRUE(shape_eq(w.shapes[0], s1));
	ASSERT_TRUE(shape_eq(w.shapes[1], s2));

	shape_free(s1);
	shape_free(s2);
	world_free(w);
}

TEST_CASE(intersect_world_with_a_ray) {
	world_t w = default_world();
	ray_t r = ray(point(0, 0, -5), vector(0, 0, 1));
	intersection_list_t ilist = world_intersect_ray(w, r);

	ASSERT_TRUE(ilist.size == 4);
	ASSERT_TRUE(req(ilist.items[0].t, 4));
	ASSERT_TRUE(req(ilist.items[1].t, 4.5));
	ASSERT_TRUE(req(ilist.items[2].t, 5.5));
	ASSERT_TRUE(req(ilist.items[3].t, 6));
	intersection_list_free(ilist);
	world_free(w);
}

TEST_CASE(shading_an_intersection) {
	world_t w = default_world();
	ray_t r = ray(point(0, 0, -5), vector(0, 0, 1));
	shape_t s = w.shapes[0];
	intersection_t i = {
		.t = 4,
		.s = s,
	};
	computations_t comps = prepare_computations(i, r);
	color_t c = world_shade_hit(w, comps);
	ASSERT_TRUE(color_eq(c, color(0.38066, 0.47583, 0.2855)));
	world_free(w);
}

TEST_CASE(shading_an_intersection_from_the_inside) {
	world_t w = default_world();
	w.point_lights[0] = point_light(point(0, 0.25, 0), color(1, 1, 1));
	ray_t r = ray(point(0, 0, 0), vector(0, 0, 1));
	shape_t s = w.shapes[1];
	intersection_t i = {
		.t = 0.5,
		.s = s,
	};
	computations_t comps = prepare_computations(i, r);
	color_t c = world_shade_hit(w, comps);
	ASSERT_TRUE(color_eq(c, color(0.90498, 0.90498, 0.90498)));
	world_free(w);
}

TEST_CASE(the_color_when_a_ray_misses) {
	world_t w = default_world();
	ray_t r = ray(point(0, 0, -5), vector(0, 1, 0));
	color_t c = world_color_at(w, r);
	ASSERT_TRUE(color_eq(c, color(0, 0, 0)));
	world_free(w);
}

TEST_CASE(the_color_when_a_ray_hits) {
	world_t w = default_world();
	ray_t r = ray(point(0, 0, -5), vector(0, 0, 1));
	color_t c = world_color_at(w, r);
	ASSERT_TRUE(color_eq(c, color(0.38066, 0.47583, 0.2855)));
	world_free(w);
}

TEST_CASE(the_color_with_an_intersection_behind_the_ray) {
	world_t w = default_world();
	w.shapes[0].material.ambient = 1;
	w.shapes[1].material.ambient = 1;
	ray_t r = ray(point(0, 0, 0.75), vector(0, 0, -1));
	color_t c = world_color_at(w, r);
	ASSERT_TRUE(color_eq(c, w.shapes[1].material.color));
	world_free(w);
}

TEST_CASE(there_is_no_shadow_when_nothing_colinear) {
	world_t w = default_world();
	vec4_t p = point(0, 10, 0);

	ASSERT_FALSE(world_point_is_shadowed(w, p, 0));
	world_free(w);
}

TEST_CASE(there_is_shadow_when_object_between) {
	world_t w = default_world();
	vec4_t p = point(10, -10, 10);

	ASSERT_TRUE(world_point_is_shadowed(w, p, 0));
	world_free(w);
}

TEST_CASE(there_is_no_shadow_when_object_is_behind_light) {
	world_t w = default_world();
	vec4_t p = point(-20, 20, -20);

	ASSERT_FALSE(world_point_is_shadowed(w, p, 0));
	world_free(w);
}

TEST_CASE(there_is_no_shadow_when_object_is_behind_point) {
	world_t w = default_world();
	vec4_t p = point(-2, 2, -2);

	ASSERT_FALSE(world_point_is_shadowed(w, p, 0));
	world_free(w);
}

TEST_CASE(shade_hit_is_given_an_intersection_in_shadow) {
	world_t w = world();
	world_add_point_light(&w, point_light(point(0, 0, -10), color(1, 1, 1)));

	shape_t s1 = sphere();
	world_add_shape(&w, s1);

	shape_t s2 = sphere();
	shape_set_transform(&s2, translation(0, 0, 10));
	world_add_shape(&w, s2);

	ray_t r = ray(point(0, 0, 5), vector(0, 0, 1));
	intersection_t i = {
		.t = 4,
		.s = s2,
	};
	computations_t comps = prepare_computations(i, r);
	color_t c = world_shade_hit(w, comps);

	ASSERT_TRUE(color_eq(c, color(0.1, 0.1, 0.1)));
	world_free(w);
}
